`get-upload-stats.py` script

jwodder opened the pull request:
Closes #83.

jwodder committed:
`get-upload-stats.py` script

yarikoptic committed:
make script executable -- has shebang

yarikoptic commented on tools/get-upload-stats.py:
```diff
@@ -0,0 +1,333 @@
+
+    def for_json(self) -> dict:
+        return {
+            "from_dt": self.from_dt.isoformat() if self.from_dt is not None else None,
```
I might have been misspecifying but no example for elderly me is provided, and it seems like it did pick up some date but just incorrect one:

<details>
<summary>my first attempt at it</summary> 

```shell
(git-annex)lena:~/proj/dandi/dandisets[draft]git
$> tools/get-upload-stats.py --from 'Mon Oct 19 23:44:25 2020 -0400' 000003           
# Changes since 2020-10-19 23:44:25-04:00

**Versions published:** 1

No changes since version 0.210812.1448 published on 2021-08-12 14:48:34-04:00

## Changes from 7263a06 to 4ae895f

* Assets added: 0 (0 bytes)
* Assets removed: 0 (0 bytes)
* Duplicates delta: 0 (0 bytes)
* Duplicates remaining: 0
* Subjects added: 0

(dev3) 1 16580 [1].....................................:Tue 21 Sep 2021 02:54:49 PM EDT:.
(git-annex)lena:~/proj/dandi/dandisets[draft]git
$> tools/get-upload-stats.py --from '2020/10/19' 000003           
Traceback (most recent call last):
  File "/home/yoh/proj/dandi/dandisets/tools/get-upload-stats.py", line 333, in <module>
    main()
  File "/home/yoh/proj/dandi/dandisets/venvs/dev3/lib/python3.9/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/home/yoh/proj/dandi/dandisets/venvs/dev3/lib/python3.9/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/home/yoh/proj/dandi/dandisets/venvs/dev3/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/yoh/proj/dandi/dandisets/venvs/dev3/lib/python3.9/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/home/yoh/proj/dandi/dandisets/tools/get-upload-stats.py", line 294, in main
    commit1, commit2 = dd.get_first_and_last_commit(from_dt, to_dt)
  File "/home/yoh/proj/dandi/dandisets/tools/get-upload-stats.py", line 177, in get_first_and_last_commit
    commits = filter_commits(commits, from_dt, to_dt)
  File "/home/yoh/proj/dandi/dandisets/tools/get-upload-stats.py", line 324, in filter_commits
    while commits and commits[0].created < from_dt:
TypeError: can't compare offset-naive and offset-aware datetimes
(dev3) 1 16581 ->1 [1].....................................:Tue 21 Sep 2021 02:55:14 PM EDT:.
(git-annex)lena:~/proj/dandi/dandisets[draft]git
$> git -C 000003 show 7263a06
commit 7263a06eebf7c0b3d2079ab2e603a1aec7dbb441
Author: DANDI Meta-user <dandi@mit.edu>
Date:   Thu May 6 23:59:46 2021 -0400

    [backups2datalad]

diff --git a/dandiset.yaml b/dandiset.yaml
index be7bd63..aaa06ba 100644
--- a/dandiset.yaml
+++ b/dandiset.yaml
@@ -1,7 +1,10 @@
 # DO NOT EDIT THIS FILE LOCALLY. ALL LOCAL UPDATES WILL BE LOST.
 # It can be edited online at https://dandiarchive.org/dandiset/000003
 # and obtained from the dandiarchive.
-about: []
+about:
+- identifier: UBERON:0002436
+  name: primary visual cortex
+  schemaKey: Anatomy
 access:
 - contactPoint:
     email: petersen.peter@gmail.com

```
</details>

<details>
<summary>and even if I set date to 2000 -- still nothing "relevant" was found</summary> 

```shell
$> tools/get-upload-stats.py --from 'Mon Oct 19 23:44:25 2000 -0400' 000003
# Changes since 2000-10-19 23:44:25-04:00

**Versions published:** 1

No changes since version 0.210812.1448 published on 2021-08-12 14:48:34-04:00

## Changes from 7263a06 to 4ae895f

* Assets added: 0 (0 bytes)
* Assets removed: 0 (0 bytes)
* Duplicates delta: 0 (0 bytes)
* Duplicates remaining: 0
* Subjects added: 0
```
</details>

Please 
- add an example of invocation to the output of `--help`
- make sure that if date parsing "didn't work out" - there is an error
- issue a warning (in logger, stderr) if history is "non-linear" (it seems what might have confused on 000003) for the relevant time period
- provide here in PR some exemplar invocations/output on some of our dandisets

jwodder replied:
@yarikoptic

> add an example of invocation to the output of `--help`

What exactly do you need the example to exhibit?  Valid date formats?  The script accepts anything that [`dateutil.parser.parse()`](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.parse) accepts.

> make sure that if date parsing "didn't work out" - there is an error

There already was an error in your second invocation, because you omitted the timezone information.  Exactly what do you want changed?

Part of the problem with your tests may be that the script (like the version-base search in `backups2datalad.py`) filters out any commits that don't contain "[backups2datalad]", which means that older commits, merge commits, and URL download commits are ignored.  Do you want the script to keep doing that or not?  Note that, if this filter is removed, then there will be a number of commit histories with the author dates out of order, as the author date wasn't always set for URL download commits.

yarikoptic replied:
> What exactly do you need the example to exhibit? Valid date formats?

yes.

> filters out any commits that don't contain "[backups2datalad]", which means that older commits, merge commits, and URL download commits are ignored. Do you want the script to keep doing that or not? 

I think that makes sense, but may be make it also consider ones which have `Ran backups2datalad.py`?

My point with above was that I got NOTHING reported really on 000003 even though I have specified a date very far in the past.  May be on some elderly dandisets like this indeed we would not be able to get sensible "report".

@satra -- which are BICCN dandisets we should primarily care about ATM, and what would be the reporting date ranges for them?

satra replied:
https://gui.dandiarchive.org/#/dandiset/search?search=biccn

satra replied:
the next due date is oct 7 i believe. but we can find the submission dates from the past as well.

yarikoptic replied:
so, next reporting period would be "beginning of the universe" -> "2021/10/07" (unlikely anyone cares about timezone)

jwodder replied:
@yarikoptic I've updated the script to take "Ran backups2datalad.py" commits into consideration, convert naïve datetimes to system time, and provide examples of acceptable timestamp formats.

satra replied:
Q2 was 7/8 and Q3 is 10/7, so that delta would be good to compute.

yarikoptic replied:
Dear @jwodder - could you please produce reports (and see if corresponds to actual changes in git history) for those dandisets for Q2 and Q3 with examples of invocation and actual reports?

jwodder replied:
@yarikoptic Assuming that the intended date ranges are Beginning of Time → July 8, 2021, and Beginning of Time → October 7, 2021, respectively, running the script on 000003 produces:

---

`./get-upload-stats.py --to '2021-07-08 00:00:00 UTC' ../000003`

# Changes up to 2021-07-08 00:00:00+00:00

**Versions published:** 0

## Changes from bbb5209 to 6b56ac0

* Assets added: 100 (2547317560305 bytes)
* Assets removed: 87 (980418516972 bytes)
* Duplicates delta: 0 (0 bytes)
* Duplicates remaining: 0
* Subjects added: 1

---

`./get-upload-stats.py --to '2021-10-07 00:00:00 UTC' ../000003`

# Changes up to 2021-10-07 00:00:00+00:00

**Versions published:** 1

No changes since version 0.210812.1448 published on 2021-08-12 14:48:45+00:00

## Changes from bbb5209 to f2da8ee

* Assets added: 100 (2547317560305 bytes)
* Assets removed: 87 (980418516972 bytes)
* Duplicates delta: 0 (0 bytes)
* Duplicates remaining: 0
* Subjects added: 1

yarikoptic replied:
For the Q3, start from next day after the end of Q2.

Please use humanize for sizes

jwodder replied:
@yarikoptic Updated outputs:

---

`./get-upload-stats.py --to '2021-07-08 00:00:00 UTC' ../000003`

# Changes up to 2021-07-08 00:00:00+00:00

**Versions published:** 0

## Changes from bbb5209 to 6b56ac0

* Assets added: 100 (2.5 TB)
* Assets removed: 87 (980.4 GB)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 1

---

`./get-upload-stats.py --from '2021-07-09 00:00:00 UTC' --to '2021-10-07 00:00:00 UTC' ../000003`

# Changes from 2021-07-09 00:00:00+00:00 to 2021-10-07 00:00:00+00:00

**Versions published:** 1

No changes since version 0.210812.1448 published on 2021-08-12 14:48:45+00:00

## Changes from ab9d37f to f2da8ee

* Assets added: 0 (0 Bytes)
* Assets removed: 0 (0 Bytes)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 0

jwodder committed:
Warn if history is nonlinear

jwodder committed:
Also accept commits with log messages containing "Ran backups2datalad.py"

jwodder committed:
Document timestamp formats and handle naïve input

jwodder committed:
Use humanize for sizes

satra commented:
@yarikoptic and @jwodder - could we update this report to include values for the columns from the spreadsheet i forwarded via email.

jwodder commented:
@satra I don't know where most of those values are found in the asset metadata.

satra commented:
@jwodder - let's go with these:

Specimen Type - from biosample.sample_type in wasderived from

the following can come from assetsSummary in dandiset metadata (and the aggregate function in dandischema should show where these are being picked up from asset metadata)
Species  -  
Modality
Technique

Anatomical Structure - presently from dandiset metadata (about key has a potential sub object called anatomy)

jwodder commented:
@satra So should the script display the number of values for each field that were changed over the time period, the value of those fields as of the last commit in the time period, or something else?

satra commented:
the script could generate a summary of state A, state B, and diff between state A and B.

jwodder commented:
@satra Is this summary supposed to be the values for each field (which doesn't sound like a "summary") or the number of different values or what?  And what exactly should the diff look like?

satra commented:
consider the summary a `set` of unique values and the difference a diff of those that indicates added/removed.

jwodder commented:
@satra 

> the following can come from assetsSummary in dandiset metadata (and the aggregate function in dandischema should show where these are being picked up from asset metadata)  
> Species -   
> Modality  
> Technique

Is modality under the "approach" key or something else?

satra commented:
> Is modality under the "approach" key or something else?

it is approach.

jwodder committed:
Summarize some metadata

jwodder commented:
@satra Metadata summary added to the output.  Here's an example:


<details>
<summary><code>./get-upload-stats.py --to '2021-07-08 00:00:00 UTC' ../000003</code></summary>

# Changes up to 2021-07-08 00:00:00+00:00

**Versions published:** 0

## bbb5209 versus 6b56ac0

* Assets added: 100 (2.5 TB)
* Assets removed: 87 (980.4 GB)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 1

### Metadata summary for bbb5209

* **Specimen Types:** [data not available]
* **Species:** [data not available]
* **Modalities:** [data not available]
* **Techniques:** [data not available]
* **Anatomical Structures:** [none]

### Metadata summary for 6b56ac0

* **Specimen Types:** [none]
* **Species:** House mouse
* **Modalities:** behavioral approach, electrophysiological approach
* **Techniques:** behavioral technique, fourier analysis technique, multi electrode extracellular electrophysiology recording technique, signal filtering technique, spike sorting technique
* **Anatomical Structures:** primary visual cortex

### Changes in metadata

* **Specimen Types:** [data not available]
* **Species:** [data not available]
* **Modalities:** [data not available]
* **Techniques:** [data not available]
* **Anatomical Structures:**:
    * Added: primary visual cortex
    * Removed: —

</details>

jwodder committed:
Typo

satra commented:
thank you @jwodder - looking better.

shouldn't the changes in metadata be more_recent_timepoint - older_timepoint?

jwodder commented:
@satra I believe it *is* that way?  Commit bbb5209 is before commit 6b56ac0.

satra commented:
> Commit bbb5209 is before commit 6b56ac0.

i would have expected the changes in metadata to show the things from `6b56ac0` rather than showing data not available.

jwodder committed:
When comparing metadata, treat unknown data in the older side as empty

jwodder commented:
@satra Adjusted.  Now, if data is not available in the earlier commit, the comparison will treat it as though there were no values for the field.  (If data is not available for the later commit, the result will still be "[data not available]".)

<details>
<summary>New sample output</summary>

# Changes up to 2021-07-08 00:00:00+00:00

**Versions published:** 0

## bbb5209 versus 6b56ac0

* Assets added: 100 (2.5 TB)
* Assets removed: 87 (980.4 GB)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 1

### Metadata summary for bbb5209

* **Specimen Types:** [data not available]
* **Species:** [data not available]
* **Modalities:** [data not available]
* **Techniques:** [data not available]
* **Anatomical Structures:** [none]

### Metadata summary for 6b56ac0

* **Specimen Types:** [none]
* **Species:** House mouse
* **Modalities:** behavioral approach, electrophysiological approach
* **Techniques:** behavioral technique, fourier analysis technique, multi electrode extracellular electrophysiology recording technique, signal filtering technique, spike sorting technique
* **Anatomical Structures:** primary visual cortex

### Changes in metadata

* **Specimen Types:** no change
* **Species:**
    * Added: House mouse
    * Removed: —
* **Modalities:**
    * Added: behavioral approach, electrophysiological approach
    * Removed: —
* **Techniques:**
    * Added: behavioral technique, fourier analysis technique, multi electrode extracellular electrophysiology recording technique, signal filtering technique, spike sorting technique
    * Removed: —
* **Anatomical Structures:**
    * Added: primary visual cortex
    * Removed: —

</details>

satra commented:
thanks looks great. i'm sure we will continue to tweak this over time with more modality and dataset specific needs. but this is a great start at the moment.

yarikoptic commented:
@jwodder instead of plain hexsha (like bbb5209) could you please use `git describe --always` so they show version of the is any

jwodder committed:
Use `git describe --always` for short_id

jwodder commented:
@yarikoptic Done.  The change doesn't make a difference to the 000003 output, as all the backups2datalad.py commits predate the single tag, so here's a different example:

<details>
<summary><code>./get-upload-stats.py ../000029</code></summary>

# Changes

**Versions published:** 6

## 840f187 versus 0.210806.2112-5-gd4c3e25

* Assets added: 5 (20.7 MB)
* Assets removed: 1 (357.8 kB)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 3

### Metadata summary for 840f187

* **Specimen Types:** [data not available]
* **Species:** [data not available]
* **Modalities:** [data not available]
* **Techniques:** [data not available]
* **Anatomical Structures:** [none]

### Metadata summary for 0.210806.2112-5-gd4c3e25

* **Specimen Types:** [none]
* **Species:** Brown rat, House mouse
* **Modalities:** behavioral approach, electrophysiological approach
* **Techniques:** behavioral technique, spike sorting technique, surgical technique
* **Anatomical Structures:** [none]

### Changes in metadata

* **Specimen Types:** no change
* **Species:**
    * Added: Brown rat, House mouse
    * Removed: —
* **Modalities:**
    * Added: behavioral approach, electrophysiological approach
    * Removed: —
* **Techniques:**
    * Added: behavioral technique, spike sorting technique, surgical technique
    * Removed: —
* **Anatomical Structures:** no change


## 0.210915.1646 versus 0.210806.2112-5-gd4c3e25

* Assets added: 0 (0 Bytes)
* Assets removed: 0 (0 Bytes)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 0

### Metadata summary for 0.210915.1646

* **Specimen Types:** [none]
* **Species:** Brown rat, House mouse
* **Modalities:** behavioral approach, electrophysiological approach
* **Techniques:** behavioral technique, spike sorting technique, surgical technique
* **Anatomical Structures:** [none]

### Metadata summary for 0.210806.2112-5-gd4c3e25

* **Specimen Types:** [none]
* **Species:** Brown rat, House mouse
* **Modalities:** behavioral approach, electrophysiological approach
* **Techniques:** behavioral technique, spike sorting technique, surgical technique
* **Anatomical Structures:** [none]

### Changes in metadata

* **Specimen Types:** no change
* **Species:** no change
* **Modalities:** no change
* **Techniques:** no change
* **Anatomical Structures:** no change

</details>

yarikoptic commented:
Looks awesome!  I wonder
- where is the date range listing is gone, says only "Changes" -- should have `[from DATE] to DATE` IMHO ideally with that `00:00:00+00:00` rstripped (if all just 0s) since just clutters 
- if we should add heading with basic dandiset information, e.g.
   - Dandiset: [DANDI:000029](https://dandiarchive.org/dandiset/000029/). [[GitHub URL](https://github.com/dandisets/000029)]
   - or could be pretty much what we show for "Cite As" (do we have it exposed in API?), e.g.
      - Last, First (2021) Test dataset for development purposes (Version 0.210915.1646) [Data set]. DANDI archive. https://doi.org/10.48324/dandi.000029/0.210915.1646
- if we should list all those versions instead of just a number.  Note that they could be hyperlinks pointing to specific version on dandiarchive AND github, e.g. in markdown it could be listing like
    - [0.210806.2112](https://dandiarchive.org/dandiset/000029/0.210806.2112).  [[GitHub URL](https://github.com/dandisets/000029/releases/tag/0.210806.2112)]

WDYT @satra?

jwodder commented:
@yarikoptic 

> where is the date range listing is gone, says only "Changes" -- should have `[from DATE] to DATE`

No "to DATE" was given on the command line, so it's not displayed.  In that case, should the header show the date of the latest commit and/or tag found, the current date & time, or something else?

jwodder committed:
Don't show zero times & timezones in timestamps

satra commented:
anything that makes it more useful and is easy (lile links, title) is welcome.

yarikoptic commented:
> No "to DATE" was given on the command line, so it's not displayed. In that case, should the header show the date of the latest commit and/or tag found, the current date & time, or something else?

I think it should show the latest date we found among commit/tags.  Current date might be misleading if we haven't fetched some fresh updates by now.

yarikoptic commented:
> anything that makes it more useful and is easy (lile links, title) is welcome.

@jwodder -- please add listing of versions and some opening to describe dandiset for which report is produced.  For now I think it is Ok to assume that we are dealing with the dandisets from our main instance, although https://github.com/dandi/dandischema/issues/76 is relevant in the long run.

jwodder committed:
Show min & max times among commits & tags in Markdown header

jwodder committed:
Show links to Dandiset and versions

jwodder commented:
@yarikoptic Done.  Here's some updated output:

<details>
<summary><code>./get-upload-stats.py ../000029</code></summary>

# Changes in [DANDI:000029](https://dandiarchive.org/dandiset/000029/) from 2019-10-08 10:24:43-04:00 to 2021-09-15 16:46:57+00:00

* **GitHub URL:** <https://github.com/dandisets/000029>
* **Versions published:**
    * [0.210712.1903](https://dandiarchive.org/dandiset/000029/0.210712.1903/) [[GitHub URL](https://github.com/dandisets/000029/releases/tag/0.210712.1903)]
    * [0.210730.1538](https://dandiarchive.org/dandiset/000029/0.210730.1538/) [[GitHub URL](https://github.com/dandisets/000029/releases/tag/0.210730.1538)]
    * [0.210805.2047](https://dandiarchive.org/dandiset/000029/0.210805.2047/) [[GitHub URL](https://github.com/dandisets/000029/releases/tag/0.210805.2047)]
    * [0.210806.1511](https://dandiarchive.org/dandiset/000029/0.210806.1511/) [[GitHub URL](https://github.com/dandisets/000029/releases/tag/0.210806.1511)]
    * [0.210806.2112](https://dandiarchive.org/dandiset/000029/0.210806.2112/) [[GitHub URL](https://github.com/dandisets/000029/releases/tag/0.210806.2112)]
    * [0.210915.1646](https://dandiarchive.org/dandiset/000029/0.210915.1646/) [[GitHub URL](https://github.com/dandisets/000029/releases/tag/0.210915.1646)]

## 840f187 versus 0.210806.2112-5-gd4c3e25

* Assets added: 5 (20.7 MB)
* Assets removed: 1 (357.8 kB)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 3

### Metadata summary for 840f187

* **Specimen Types:** [data not available]
* **Species:** [data not available]
* **Modalities:** [data not available]
* **Techniques:** [data not available]
* **Anatomical Structures:** [none]

### Metadata summary for 0.210806.2112-5-gd4c3e25

* **Specimen Types:** [none]
* **Species:** Brown rat, House mouse
* **Modalities:** behavioral approach, electrophysiological approach
* **Techniques:** behavioral technique, spike sorting technique, surgical technique
* **Anatomical Structures:** [none]

### Changes in metadata

* **Specimen Types:** no change
* **Species:**
    * Added: Brown rat, House mouse
    * Removed: —
* **Modalities:**
    * Added: behavioral approach, electrophysiological approach
    * Removed: —
* **Techniques:**
    * Added: behavioral technique, spike sorting technique, surgical technique
    * Removed: —
* **Anatomical Structures:** no change


## 0.210915.1646 versus 0.210806.2112-5-gd4c3e25

* Assets added: 0 (0 Bytes)
* Assets removed: 0 (0 Bytes)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 0

### Metadata summary for 0.210915.1646

* **Specimen Types:** [none]
* **Species:** Brown rat, House mouse
* **Modalities:** behavioral approach, electrophysiological approach
* **Techniques:** behavioral technique, spike sorting technique, surgical technique
* **Anatomical Structures:** [none]

### Metadata summary for 0.210806.2112-5-gd4c3e25

* **Specimen Types:** [none]
* **Species:** Brown rat, House mouse
* **Modalities:** behavioral approach, electrophysiological approach
* **Techniques:** behavioral technique, spike sorting technique, surgical technique
* **Anatomical Structures:** [none]

### Changes in metadata

* **Specimen Types:** no change
* **Species:** no change
* **Modalities:** no change
* **Techniques:** no change
* **Anatomical Structures:** no change

</details>

yarikoptic commented on tools/get-upload-stats.py:
```diff
@@ -0,0 +1,580 @@
+                )
+            )
+        keys1, keys2 = key_qtys
+        added_keys = keys2.keys() - keys1.keys()
```
note to myself: damn -- didn't know `dict.keys()` can do that.

jwodder replied:
@yarikoptic It's because `dict.keys()` returns a [dictionary view](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects), which is set-like, so you can do set operations on it.

yarikoptic commented:
Looks great to me!  doing self-educational code review pass now.
Would you be so kind to sweep through all dandisets (e.g. for the past 4 months interval) and dump that output somewhere on drogon, while also making sure that script doesn't error out.

yarikoptic commented on tools/get-upload-stats.py:
```diff
@@ -0,0 +1,580 @@
+            if info.get("type") == "file" and relpath.parts[0] not in IGNORED:
+                path = str(PurePosixPath(relpath))
+                subject: Optional[str]
+                if m := re.fullmatch(r"sub-([^/]+)/[^/]+\.nwb", path):
```
I don't think we should limit to .nwb files here (we have .tif etc now), so I guess smth like 
```suggestion
                if m := re.match(r"sub-([^/]+)/[^/]+", path):
```
?

@satra WDYT -- is it worth "ignoring" .json side car files? I would say - we should not bother (i.e. we should count them since they are also assets ATM and likely in the future)

jwodder replied:
@yarikoptic Note that the limiting to NWB files only applies to extracting the subject from the asset path; aside from calculating the subject, the script still takes into account all files other than `.dandi/`, `.datalad/`, `.gitattributes`, and `dandiset.yaml`.

yarikoptic replied:
But we have dandisets with subjects and without .nwb files - I still don't see how it is nwb specific

yarikoptic reviewed (changes requested):
glorious.  Just one change request: we should not limit to .nwb files

jwodder committed:
Handle some odd metadata

jwodder committed:
Extract subjects from all file types, not just NWB

yarikoptic commented:
Dear @jwodder , please ping whenever the sweep is complete and you expect no further changes -- we could as well proceed and merge I think.

jwodder commented:
@yarikoptic Script output (run with `--from '2021-06-01 00:00:00+00:00'`) is now stored on drogon at `/home/jwodder/dandiset-stats`.  I had to skip the following datasets, as they had either zero or one commits in the given timeframe:

* 000062
* 000123
* 000131
* 000132
* 000133
* 000134
* 000135
* 000136
* 000137
* 000138
* 000141
* 000142

yarikoptic commented:
got confused by the one for 000126 -- has 
```
## e63967c versus 0.210813.0327
...
## 0.210813.0327 versus 0.210813.0327
...
```
the idea is to show differences between each release versions?

I also wondered if there is no change of any kind for any of the fields, should we just have short "no change" instead of a detailed list with "no change" (or 0s) for each of the entries?

jwodder commented:
@yarikoptic For 000126, that happened because the latest commit is also the latest tagged commit.

jwodder committed:
Show "No change" if there were no changes in metadata

jwodder committed:
Bugfix

jwodder commented:
@yarikoptic That was a bug.  I've regenerated the output files, and 000126.md now shows "No changes since version 0.210813.0327 published on 2021-08-13 03:27:03+00:00".  Also, if there were no changes in metadata, the "Changes in metadata" section will now read "No change".

yarikoptic commented:
> I had to skip the following datasets, as they had either zero or one commits in the given timeframe:
> 
> * 000062

for such datasets, report should still be produced, just state that there were no changes

jwodder commented:
@yarikoptic For datasets with zero commits in the timeframe, there's nothing to extract the Dandiset ID or implied timestamp bounds from.  How should that be handled?

yarikoptic commented:
Dandiset ID should be extracted regardless of the range/commits from `dandiset.yaml` in the current/latest state.

time range: just state that in that time range (as specified/provided, omitting `from DATE` if not specified) no commits (or "no changes") detected.

jwodder committed:
Produce reports for datasets with insufficient commits

jwodder commented:
@yarikoptic Code updated and output files regenerated.

yarikoptic commented:
> Also, if there were no changes in metadata, the "Changes in metadata" section will now read "No change".

000126 is an example of that.

But what about replacing those 
```
* **Specimen Types:** [none]
* **Species:** [none]
* **Modalities:** [none]
* **Techniques:** [none]
* **Anatomical Structures:** [none]

```
and 
```
* Assets added: 0 (0 Bytes)
* Assets removed: 0 (0 Bytes)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 0
```
also with "No change" ?

yarikoptic commented:
oh, the metadata ones are not "change" but rather the description of the state.... ok - let's leave that one as is, but let's replace the size differences of 0s with just "no change".

jwodder commented:
@yarikoptic Do you mean that you want this:

```
* Assets added: no change
* Assets removed: no change
* Duplicates delta: no change
* Duplicates remaining: 0
* Subjects added: 0
```

or this:

```
* Assets: no change
* Duplicates: no change
* Duplicates remaining: 0
* Subjects added: 0
```

or something else?

yarikoptic commented:
no, just

```
No change
```

jwodder commented:
@yarikoptic But the "Duplicates remaining" line isn't about changes.  Should this line always be included, or should it be ignored when determining whether there were any changes?

yarikoptic commented:
I am ok either way (e.g. detailed listing if some duplicates remain, or just have "no changes" if duplicates remain -- after all, the same number of duplicates would be if there is "no change" -- so may be ignoring duplicates remaining would be "most consistent" with such argument)

jwodder committed:
Show "No change in assets" on no change in assets

jwodder committed:
Remove blank line from end of Markdown output

jwodder commented:
@yarikoptic Updated and regenerated.

yarikoptic reviewed (approved):
Looks great! Thank you @jwodder!
tests seems to start failing for backups2datalad.py but that is a separate issue

```
=========================== short test summary info ============================
FAILED test_backups2datalad.py::test_1 - ValueError: Dandiset 100335 is Pendi...
FAILED test_backups2datalad.py::test_2 - ValueError: Dandiset 100337 is Pendi...
```
so please merge this PR @jwodder if you agree that fails are unrelated!

jwodder commented:
@yarikoptic I think the failures may have to do with the recent changes to dandi-api's validation routines.  It's definitely not related to this PR.
